### DISCUSSION TITLE
Enums & Interfaces

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#208] [BC Best Practices]
I try to put "the use of interfaces" into a guideline.  Here's one suggestion:

- When we create an extensible enum, shouldn't it be (in general) accompanied with an interface
- Whenever you need a case-statement for an extensible enum (I would even argue that it doesn't matter if the enum is extensible or not), shouldn't you have used an interface instead?

Here's a thread and poll on Twitter: https://twitter.com/waldo1001/status/1600160428205953025?s=20&t=sEYmkMuf31LOvxakj2zy1Q

**Comment by tscottjendev** [comment]
Another question, if you have an enum as a field in a table.  Should you have a interface "accessor" function to return the interface?  This would stop having to have an interface variable and then assign the enum to the interface variable.  
If yes, what should the function name be called?

so this
```al
table 50000 SomeTable {
   fields{
       field(1; PK; Code[10]) {}
       field(2; EnumField; Enum SomeEnum) {}
    }

    procedure EnumFieldInterace(): Interface SomeInterface  //what should the procedure name be?
    begin
        exit(EnumField);
    end;
}

codeunit 50000 SomeCodeunit {
procedure DoSomething()
var
    SomeTable: Record SomeTable;
begin
   SomeTable.Get();
   SomeTable.EnumFieldInterface().InterfaceMethod();
end;
}
```
rather than
```al
table 50000 SomeTable {
   fields{
       field(1; PK; Code[10]) {}
       field(2; EnumField; Enum SomeEnum) {}
    }
}

codeunit 50000 SomeCodeunit {
procedure DoSomething()
var
    SomeTable: record SomeTable;
    SomeInterface: interface SomeInterface;
begin
    SomeTable.Get();
    SomeInterface := SomeTable.EnumField
   SomeInterface.InterfaceMethod();
end;
}
```

**Comment by JeremyVyska** [comment]
I saw some really interesting stuff from a couple of wise guys today at TechDays  that should be worked into this.
